File: backend/internal/storage/s3.go
package storage

import (
	"context"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"

	"github.com/imranp/kerf/backend/internal/config"
)

// S3 is an S3 (or S3-compatible) Storage implementation.
type S3 struct {
	client    *s3.Client
	presigner *s3.PresignClient
	bucket    string
	publicURL string // S3PublicURLBase (e.g. virtual-hosted bucket origin)
	cdnURL    string // CDNBaseURL (bunny.net Pull Zone, CloudFront, etc.)

	// In-flight multipart uploads, keyed by upload session id.
	//
	// NON-DURABLE: this lives only in process memory. A server restart
	// abandons in-flight uploads (the client has to start over from chunk
	// 0). Acceptable for v1 — moving this into Redis or DB is straightforward
	// once it becomes painful.
	mpMu      sync.Mutex
	multipart map[string]*s3MultipartState
}

type s3MultipartState struct {
	// uploadID is the value returned by S3's CreateMultipartUpload.
	uploadID string
	// dstKey is the eventual target key (where ConcatChunksTo will deliver
	// the assembled object). Captured at multipart-create time so the
	// frontend doesn't have to repeat it on every chunk.
	dstKey string
	// parts is the list of completed parts so far, keyed by index. S3
	// expects these in part-number order on CompleteMultipartUpload.
	parts map[int]s3types.CompletedPart
}

// NewS3 builds a Storage backed by S3.
func NewS3(cfg *config.Config) (*S3, error) {
	ctx := context.Background()
	loaders := []func(*awsconfig.LoadOptions) error{}
	if cfg.S3Region != "" {
		loaders = append(loaders, awsconfig.WithRegion(cfg.S3Region))
	}
	if cfg.S3AccessKeyID != "" && cfg.S3SecretAccessKey != "" {
		loaders = append(loaders, awsconfig.WithCredentialsProvider(
			credentials.NewStaticCredentialsProvider(cfg.S3AccessKeyID, cfg.S3SecretAccessKey, ""),
		))
	}
	awsCfg, err := awsconfig.LoadDefaultConfig(ctx, loaders...)
	if err != nil {
		return nil, fmt.Errorf("storage/s3: load config: %w", err)
	}

	clientOpts := []func(*s3.Options){}
	if cfg.S3Endpoint != "" {
		ep := cfg.S3Endpoint
		clientOpts = append(clientOpts, func(o *s3.Options) {
			o.BaseEndpoint = aws.String(ep)
			o.UsePathStyle = true
		})
	}
	client := s3.NewFromConfig(awsCfg, clientOpts...)
	presigner := s3.NewPresignClient(client)

	return &S3{
		client:    client,
		presigner: presigner,
		bucket:    cfg.S3Bucket,
		publicURL: strings.TrimRight(cfg.S3PublicURLBase, "/"),
		cdnURL:    strings.TrimRight(cfg.CDNBaseURL, "/"),
		multipart: make(map[string]*s3MultipartState),
	}, nil
}

// Put uploads an object.
func (s *S3) Put(ctx context.Context, key string, body io.Reader, contentType string, size int64) (PutResult, error) {
	if contentType == "" {
		contentType = guessContentType(key)
	}
	in := &s3.PutObjectInput{
		Bucket:      aws.String(s.bucket),
		Key:         aws.String(key),
		Body:        body,
		ContentType: aws.String(contentType),
	}
	if size > 0 {
		in.ContentLength = aws.Int64(size)
	}
	if _, err := s.client.PutObject(ctx, in); err != nil {
		return PutResult{}, fmt.Errorf("storage/s3: put: %w", err)
	}
	return PutResult{Key: key, Size: size, ContentType: contentType}, nil
}

// Get streams an object back from S3.
func (s *S3) Get(ctx context.Context, key string) (io.ReadCloser, string, error) {
	out, err := s.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return nil, "", fmt.Errorf("storage/s3: get: %w", err)
	}
	ct := ""
	if out.ContentType != nil {
		ct = *out.ContentType
	} else {
		ct = guessContentType(key)
	}
	return out.Body, ct, nil
}

// Delete removes an object.
func (s *S3) Delete(ctx context.Context, key string) error {
	_, err := s.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return fmt.Errorf("storage/s3: delete: %w", err)
	}
	return nil
}

// SignedURL returns a presigned GET URL for the object.
func (s *S3) SignedURL(ctx context.Context, key string, ttl time.Duration) (string, error) {
	if ttl <= 0 {
		ttl = 15 * time.Minute
	}
	req, err := s.presigner.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(key),
	}, s3.WithPresignExpires(ttl))
	if err != nil {
		return "", err
	}
	return req.URL, nil
}

// PublicURL prefers a configured CDN base (bunny.net Pull Zone, CloudFront,
// etc.) over the bucket's own public URL. When neither is set we fall back
// to a virtual-hosted-style S3 URL — which only works if the bucket itself
// is publicly readable. Most cloud deploys should set CDNBaseURL.
//
// Cache-buster: when updatedAt is non-zero we append "?v=<unix>" so a
// fresh upload busts CDN edge caches without an explicit purge.
func (s *S3) PublicURL(key string, updatedAt time.Time) string {
	base := ""
	switch {
	case s.cdnURL != "":
		base = s.cdnURL + "/" + escapeKey(key)
	case s.publicURL != "":
		base = s.publicURL + "/" + escapeKey(key)
	default:
		base = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", s.bucket, escapeKey(key))
	}
	if !updatedAt.IsZero() {
		return base + "?v=" + strconv.FormatInt(updatedAt.Unix(), 10)
	}
	return base
}

// --- Chunked upload helpers ------------------------------------------------
//
// We map the abstract chunked-upload API onto S3's native multipart upload.
// The first PutChunk call lazily kicks off CreateMultipartUpload using a
// temp object key under `_uploads/<uploadKey>`; subsequent PutChunks call
// UploadPart. ConcatChunksTo calls CompleteMultipartUpload, which copies the
// assembled object to the requested final key (S3 semantics — the multipart
// target IS the final key, so we issue a server-side copy after completion
// to land it at dstKey, then delete the temp). DeleteUpload calls
// AbortMultipartUpload.

func (s *S3) tempKey(uploadKey string) string {
	return "_uploads/" + uploadKey
}

func (s *S3) ensureMultipart(ctx context.Context, uploadKey string) (*s3MultipartState, error) {
	s.mpMu.Lock()
	defer s.mpMu.Unlock()
	if st, ok := s.multipart[uploadKey]; ok {
		return st, nil
	}
	tempKey := s.tempKey(uploadKey)
	out, err := s.client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(tempKey),
	})
	if err != nil {
		return nil, fmt.Errorf("storage/s3: create multipart: %w", err)
	}
	st := &s3MultipartState{
		uploadID: aws.ToString(out.UploadId),
		dstKey:   tempKey,
		parts:    make(map[int]s3types.CompletedPart),
	}
	s.multipart[uploadKey] = st
	return st, nil
}

func (s *S3) lookupMultipart(uploadKey string) *s3MultipartState {
	s.mpMu.Lock()
	defer s.mpMu.Unlock()
	return s.multipart[uploadKey]
}

func (s *S3) forgetMultipart(uploadKey string) {
	s.mpMu.Lock()
	defer s.mpMu.Unlock()
	delete(s.multipart, uploadKey)
}

// PutChunk uploads chunk #chunkIndex as an S3 part. Part numbers are 1-based
// in S3 — we add one to the 0-indexed chunk index on the wire.
func (s *S3) PutChunk(ctx context.Context, uploadKey string, chunkIndex int, body io.Reader) error {
	if chunkIndex < 0 {
		return fmt.Errorf("storage/s3: negative chunk index")
	}
	st, err := s.ensureMultipart(ctx, uploadKey)
	if err != nil {
		return err
	}
	partNumber := int32(chunkIndex + 1)
	out, err := s.client.UploadPart(ctx, &s3.UploadPartInput{
		Bucket:     aws.String(s.bucket),
		Key:        aws.String(st.dstKey),
		UploadId:   aws.String(st.uploadID),
		PartNumber: aws.Int32(partNumber),
		Body:       body,
	})
	if err != nil {
		return fmt.Errorf("storage/s3: upload part %d: %w", chunkIndex, err)
	}
	s.mpMu.Lock()
	st.parts[chunkIndex] = s3types.CompletedPart{
		ETag:       out.ETag,
		PartNumber: aws.Int32(partNumber),
	}
	s.mpMu.Unlock()
	return nil
}

// ListChunks returns the chunk indices we've recorded as uploaded. Pulled
// from the in-memory part map.
func (s *S3) ListChunks(ctx context.Context, uploadKey string) ([]int, error) {
	st := s.lookupMultipart(uploadKey)
	if st == nil {
		return []int{}, nil
	}
	s.mpMu.Lock()
	out := make([]int, 0, len(st.parts))
	for idx := range st.parts {
		out = append(out, idx)
	}
	s.mpMu.Unlock()
	sort.Ints(out)
	return out, nil
}

// ConcatChunksTo finishes the multipart upload, then copies the assembled
// object to the requested destination key (since multipart writes to a fixed
// key chosen at create time). Returns the total byte count via HeadObject.
func (s *S3) ConcatChunksTo(ctx context.Context, uploadKey string, dstKey string) (int64, error) {
	st := s.lookupMultipart(uploadKey)
	if st == nil {
		return 0, fmt.Errorf("storage/s3: no multipart state for upload %q (server restart?)", uploadKey)
	}
	s.mpMu.Lock()
	indices := make([]int, 0, len(st.parts))
	for idx := range st.parts {
		indices = append(indices, idx)
	}
	sort.Ints(indices)
	parts := make([]s3types.CompletedPart, 0, len(indices))
	for _, idx := range indices {
		parts = append(parts, st.parts[idx])
	}
	tempKey := st.dstKey
	uploadID := st.uploadID
	s.mpMu.Unlock()

	if len(parts) == 0 {
		return 0, fmt.Errorf("storage/s3: no parts uploaded for %q", uploadKey)
	}

	if _, err := s.client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
		Bucket:          aws.String(s.bucket),
		Key:             aws.String(tempKey),
		UploadId:        aws.String(uploadID),
		MultipartUpload: &s3types.CompletedMultipartUpload{Parts: parts},
	}); err != nil {
		return 0, fmt.Errorf("storage/s3: complete multipart: %w", err)
	}

	// Server-side copy to the final key, then delete the temp.
	copySource := s.bucket + "/" + tempKey
	if _, err := s.client.CopyObject(ctx, &s3.CopyObjectInput{
		Bucket:     aws.String(s.bucket),
		Key:        aws.String(dstKey),
		CopySource: aws.String(copySource),
	}); err != nil {
		return 0, fmt.Errorf("storage/s3: copy assembled object: %w", err)
	}
	_, _ = s.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(tempKey),
	})

	// Pull final size via HeadObject — caller relies on this for the files row.
	head, err := s.client.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(dstKey),
	})
	if err != nil {
		return 0, fmt.Errorf("storage/s3: head assembled object: %w", err)
	}
	s.forgetMultipart(uploadKey)
	if head.ContentLength != nil {
		return *head.ContentLength, nil
	}
	return 0, nil
}

// DeleteUpload aborts the multipart upload (if any). Idempotent.
func (s *S3) DeleteUpload(ctx context.Context, uploadKey string) error {
	st := s.lookupMultipart(uploadKey)
	if st == nil {
		return nil
	}
	_, err := s.client.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
		Bucket:   aws.String(s.bucket),
		Key:      aws.String(st.dstKey),
		UploadId: aws.String(st.uploadID),
	})
	s.forgetMultipart(uploadKey)
	if err != nil {
		// Don't surface AWS errors here — the goal of DeleteUpload is to
		// best-effort wipe state. The S3 lifecycle rule will eventually
		// reap any orphans.
		return fmt.Errorf("storage/s3: abort multipart: %w", err)
	}
	return nil
}

File: kerf-core/src/kerf_core/storage/s3.py
import asyncio
import boto3
import functools
import io
import logging
import mimetypes
from datetime import datetime
from typing import IO

from botocore.config import Config as BotoConfig

from .base import PutResult, Storage


async def _run_sync(func, /, *args, **kwargs):
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(None, functools.partial(func, *args, **kwargs))

logger = logging.getLogger(__name__)

CHUNK_DIR = "_uploads"


class S3Storage(Storage):
    def __init__(
        self,
        bucket: str,
        region: str = "",
        access_key_id: str = "",
        secret_access_key: str = "",
        endpoint: str = "",
        public_url_base: str = "",
        cdn_url: str = "",
    ):
        self.bucket = bucket
        self.public_url_base = public_url_base.rstrip("/") if public_url_base else ""
        self.cdn_url = cdn_url.rstrip("/") if cdn_url else ""

        client_kwargs = {"region_name": region} if region else {}

        if access_key_id and secret_access_key:
            client_kwargs["aws_access_key_id"] = access_key_id
            client_kwargs["aws_secret_access_key"] = secret_access_key

        if endpoint:
            client_kwargs["endpoint_url"] = endpoint
            config = boto3.session.Config(s3={"addressing_style": "path"})
            client_kwargs["config"] = config

        self.client = boto3.client("s3", **client_kwargs)
        self._multipart: dict[str, dict] = {}

    def _temp_key(self, upload_key: str) -> str:
        return f"{CHUNK_DIR}/{upload_key}"

    async def put(
        self, key: str, body: IO[bytes], content_type: str, size: int
    ) -> PutResult:
        if not content_type:
            content_type = self._guess_content_type(key)

        content = await _run_sync(body.read) if hasattr(body, "read") else body

        put_kwargs = dict(
            Bucket=self.bucket,
            Key=key,
            Body=content,
            ContentType=content_type,
        )
        if size > 0:
            put_kwargs["ContentLength"] = size

        await _run_sync(self.client.put_object, **put_kwargs)

        return PutResult(key=key, size=size, content_type=content_type)

    async def get(self, key: str) -> tuple[io.BytesIO, str]:
        response = await _run_sync(self.client.get_object, Bucket=self.bucket, Key=key)
        body = await _run_sync(response["Body"].read)
        content_type = response.get("ContentType", self._guess_content_type(key))
        return io.BytesIO(body), content_type

    async def delete(self, key: str) -> None:
        await _run_sync(self.client.delete_object, Bucket=self.bucket, Key=key)

    async def signed_url(self, key: str, ttl_seconds: int = 900) -> str:
        if ttl_seconds <= 0:
            ttl_seconds = 900

        return await _run_sync(
            self.client.generate_presigned_url,
            "get_object",
            Params={"Bucket": self.bucket, "Key": key},
            ExpiresIn=ttl_seconds,
        )

    def public_url(self, key: str, updated_at: datetime | None = None) -> str:
        if self.cdn_url:
            base = f"{self.cdn_url}/{self._escape_key(key)}"
        elif self.public_url_base:
            base = f"{self.public_url_base}/{self._escape_key(key)}"
        else:
            base = f"https://{self.bucket}.s3.amazonaws.com/{self._escape_key(key)}"

        if updated_at:
            base += f"?v={int(updated_at.timestamp())}"
        return base

    async def put_chunk(
        self, upload_key: str, chunk_index: int, body: IO[bytes]
    ) -> None:
        if chunk_index < 0:
            raise ValueError("Negative chunk index")

        await self._ensure_multipart(upload_key)

        state = self._multipart[upload_key]
        content = body.read() if hasattr(body, "read") else body

        part_number = chunk_index + 1
        response = await _run_sync(
            self.client.upload_part,
            Bucket=self.bucket,
            Key=state["dst_key"],
            UploadId=state["upload_id"],
            PartNumber=part_number,
            Body=content,
        )

        state["parts"][chunk_index] = {
            "ETag": response["ETag"],
            "PartNumber": part_number,
        }

    async def list_chunks(self, upload_key: str) -> list[int]:
        state = self._multipart.get(upload_key)
        if not state:
            return []
        return sorted(state["parts"].keys())

    async def concat_chunks_to(self, upload_key: str, dst_key: str) -> int:
        state = self._multipart.get(upload_key)
        if not state:
            raise ValueError(f"No multipart state for upload {upload_key}")

        indices = sorted(state["parts"].keys())
        if not indices:
            raise ValueError(f"No parts uploaded for {upload_key}")

        parts = [state["parts"][idx] for idx in indices]
        temp_key = state["dst_key"]
        upload_id = state["upload_id"]

        await _run_sync(
            self.client.complete_multipart_upload,
            Bucket=self.bucket,
            Key=temp_key,
            UploadId=upload_id,
            MultipartUpload={"Parts": parts},
        )

        copy_source = {"Bucket": self.bucket, "Key": temp_key}
        await _run_sync(
            self.client.copy_object,
            Bucket=self.bucket,
            Key=dst_key,
            CopySource=copy_source,
        )
        await _run_sync(self.client.delete_object, Bucket=self.bucket, Key=temp_key)

        head = await _run_sync(self.client.head_object, Bucket=self.bucket, Key=dst_key)
        size = head.get("ContentLength", 0)

        del self._multipart[upload_key]
        return size

    async def delete_upload(self, upload_key: str) -> None:
        state = self._multipart.get(upload_key)
        if not state:
            return

        try:
            await _run_sync(
                self.client.abort_multipart_upload,
                Bucket=self.bucket,
                Key=state["dst_key"],
                UploadId=state["upload_id"],
            )
        except Exception:
            pass

        del self._multipart[upload_key]

    async def _ensure_multipart(self, upload_key: str) -> None:
        if upload_key in self._multipart:
            return

        temp_key = self._temp_key(upload_key)
        response = await _run_sync(
            self.client.create_multipart_upload, Bucket=self.bucket, Key=temp_key
        )

        self._multipart[upload_key] = {
            "upload_id": response["UploadId"],
            "dst_key": temp_key,
            "parts": {},
        }

    def _escape_key(self, key: str) -> str:
        import urllib.parse

        parts = key.strip("/").split("/")
        return "/".join(urllib.parse.quote(p, safe="") for p in parts)

    def _guess_content_type(self, key: str) -> str:
        ext = key.split(".")[-1].lower() if "." in key else ""
        if ext in ("step", "stp"):
            return "model/step"
        ct, _ = mimetypes.guess_type(key)
        return ct or "application/octet-stream"
